src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs

```diff
@@ -12,12 +12,11 @@ namespace System.Text.Json.Serialization.Converters
     /// representing the dictionary element key and value.
     /// </summary>
     internal sealed class IDictionaryConverter<TCollection>
-        : DictionaryDefaultConverter<TCollection, object?>
+        : DictionaryDefaultConverter<TCollection, string, object?>
         where TCollection : IDictionary
     {
-        protected override void Add(object? value, JsonSerializerOptions options, ref ReadStack state)
+        protected override void Add(string key, object? value, JsonSerializerOptions options, ref ReadStack state)
```

Comment:
base method signature is generic, it is not possible to pass an incompatible type to this method. 

```cs
protected abstract void Add(TKey key, TValue value, JsonSerializerOptions options, ref ReadStack state)
```

In case the JSON property name fails to cast to `TKey` we would throw on the key converter.